rfd/0203-database-healthchecks.md

```diff
@@ -0,0 +1,606 @@
+      BoolValue Enabled = 1 [
+          (gogoproto.nullable) = true,
+          (gogoproto.jsontag) = "enabled,omitempty",
+          (gogoproto.customtype) = "BoolOption"
+      ];
+      // Timeout is the health check connection establishment timeout.
+      // An attempt that times out is a failed attempt.
+      int64 Timeout = 2 [
+        (gogoproto.jsontag) = "timeout,omitempty",
+        (gogoproto.casttype) = "Duration"
+      ];
+      // Interval is the time between each health check.
+      int64 Interval = 3 [
+        (gogoproto.jsontag) = "interval,omitempty",
+        (gogoproto.casttype) = "Duration"
+      ];
+      // HealthyThreshold is the number of consecutive passing health checks after
+      // which a target's health status becomes "healthy".
+      uint32 HealthyThreshold = 4 [(gogoproto.jsontag) = "healthy_threshold,omitempty"];
+      // UnhealthyThreshold is the number of consecutive failing health checks after
+      // which a target's health status becomes "unhealthy".
+      uint32 UnhealthyThreshold = 5 [(gogoproto.jsontag) = "unhealthy_threshold,omitempty"];
+    }
+
+A new message, TargetHealth, will be added as a field named `target_health` in DatabaseServerSpecV3.
+
+    // TargetHealth describes the health status of network connectivity between
+    // an agent and a resource.
+    message TargetHealth {
+      // Addr is the target address.
+      string Addr = 1 [(gogoproto.jsontag) = "addr,omitempty"];
+      // Protocol is the health check protocol such as "tcp".
+      string Protocol = 2 [(gogoproto.jsontag) = "protocol,omitempty"];
+      // Status is the health status, one of "", "init", "healthy", "unhealthy".
+      string Status = 3 [(gogoproto.jsontag) = "status,omitempty"];
+      // TransitionTimestamp is the time that the last status transition occurred.
+      google.protobuf.Timestamp TransitionTimestamp = 4 [
+        (gogoproto.jsontag) = "transition_timestamp,omitempty",
+        (gogoproto.stdtime) = true,
+        (gogoproto.nullable) = true
+      ];
+      // TransitionReason explains why the last transition occurred.
+      string TransitionReason = 5 [(gogoproto.jsontag) = "transition_reason,omitempty"];
+      // TransitionError shows the health check error observed when the transition
+      // happened. Empty when transitioning to "healthy".
+      string TransitionError = 6 [(gogoproto.jsontag) = "transition_error,omitempty"];
+      // Message is additional information meant for a user.
+      string Message = 7 [(gogoproto.jsontag) = "message,omitempty"];
+    }
+
+### Backward Compatibility
+
+Older database agents will not perform health checks, so they will always report health status `""` (unknown), which is equivalent to agents that have disabled health checks.
+
+An unknown health status will not prevent connections to that database through that agent, but it will prioritize `healthy` agents first.
+
+### Audit Events
+
+No new audit events will be added.
```

Comment:
Thunk it would make sense to generate audit events when a resource transitions from healthy to unhealthy and vice versa?

Reply:
It should not have the same backend pressure concerns because audit events are stored in a different backend than cluster data and the rest of cluster activity will generate many more audit events for these to make any sort of impact but no objections with leaving this out of initial scope.